hw/drivers/bq27z561/src/bq27z561.c

```diff
@@ -587,6 +587,104 @@ bq27x561_rd_flash(struct bq27z561 *dev, uint16_t addr, uint8_t *buf, int buflen)
     return ret;
 }
 
+bq27z561_err_t
+bq27z561_wr_flash(struct bq27z561 *dev, uint16_t addr, uint8_t *buf, int buflen)
+{
+    uint8_t tmpbuf[BQ27Z561_MAX_FLASH_RW_LEN + 2];
+    uint8_t chksum;
+    bq27z561_err_t ret;
+    int rc;
+#if MYNEWT_VAL(BUS_DRIVER_PRESENT)
+    struct os_dev *odev = &dev->dev.dev;
+#else
+    struct hal_i2c_master_data i2c;
+#endif
+
+    if ((buflen == 0) || (!buf) || (buflen > BQ27Z561_MAX_FLASH_RW_LEN)) {
+        return BQ27Z561_ERR_INV_PARAMS;
+    }
+
+    if ((addr < BQ27Z561_FLASH_BEG_ADDR) || (addr > BQ27Z561_FLASH_END_ADDR)) {
+        return BQ27Z561_ERR_INV_FLASH_ADDR;
+    }
+
+#if MYNEWT_VAL(BUS_DRIVER_PRESENT)
+    rc = bus_node_lock(odev, BUS_NODE_LOCK_DEFAULT_TIMEOUT);
+    if (rc) {
+        ret = BQ27Z561_ERR_I2C_ERR;
+        goto err;
+    }
+#else
+    i2c.address = dev->bq27_itf.itf_addr;
+
+    rc = bq27z561_itf_lock(&dev->bq27_itf, MYNEWT_VAL(BQ27Z561_ITF_LOCK_TMO));
+    if (rc) {
+        return rc;
+    }
+#endif
+
+    tmpbuf[0] = BQ27Z561_REG_MFRG_ACC;
+    tmpbuf[1] = (uint8_t)addr;
+    tmpbuf[2] = (uint8_t)(addr >> 8);
+    memcpy(&tmpbuf[3], buf, buflen);
+
+#if MYNEWT_VAL(BUS_DRIVER_PRESENT)
+    rc = bus_node_simple_write(odev, tmpbuf, buflen + 3);
+    if (rc) {
+        (void)bus_node_unlock(odev);
+        ret = BQ27Z561_ERR_I2C_ERR;
+        goto err;
+    }
+#else
+    i2c.len = buflen + 3;
+    i2c.buffer = tmpbuf;
+    rc = i2cn_master_write(dev->bq27_itf.itf_num, &i2c, MYNEWT_VAL(BQ27Z561_I2C_TIMEOUT_TICKS), 1,
+                           MYNEWT_VAL(BQ27Z561_I2C_RETRIES));
+    if (rc != 0) {
+        BQ27Z561_LOG(ERROR, "I2C reg read (wr) failed 0x%02X\n", tmpbuf[0]);
+        ret = BQ27Z561_ERR_I2C_ERR;
+        goto err;
+    }
+#endif
+
+    /* Calculate checksum */
+    chksum = bq27z561_calc_chksum(&tmpbuf[1], buflen + 2);
+
+    /* Write checksum and length */
+    tmpbuf[0] = BQ27Z561_REG_CHKSUM;
+    tmpbuf[1] = chksum;
+    tmpbuf[2] = buflen + 4;
+
+#if MYNEWT_VAL(BUS_DRIVER_PRESENT)
+    rc = bus_node_simple_write(odev, tmpbuf, 3);
+    if (rc) {
+        (void)bus_node_unlock(odev);
+        ret = BQ27Z561_ERR_I2C_ERR;
+        goto err;
+    }
+
+    (void)bus_node_unlock(odev);
+
+    ret = BQ27Z561_OK;
+
+err:
+#else
+    i2c.len = 3;
+    i2c.buffer = tmpbuf;
+    rc = i2cn_master_write(dev->bq27_itf.itf_num, &i2c, MYNEWT_VAL(BQ27Z561_I2C_TIMEOUT_TICKS), 1,
+                           MYNEWT_VAL(BQ27Z561_I2C_RETRIES));
+    if (rc != 0) {
+        BQ27Z561_LOG(ERROR, "I2C reg read (wr) failed 0x%02X\n", tmpbuf[0]);
+        ret = BQ27Z561_ERR_I2C_ERR;
+    }
+
+err:
+    bq27z561_itf_unlock(&dev->bq27_itf);
```

Comment:
Calls to unlock device in error path when BUS_DRIVER_PRESENT happen before goto err;. Calls to unlock device in error path when BUS_DRIVER_PRESENT is not set happen after err: label.

I would suggest making the logic to be the same for both cases. My preference would be to do the unlocking after the err: label.

Reply:
Agreed. Updated.